File: BinaryTreeInorderTraversal.java

// Time Complexity O(n) Space Complexity O(n)
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        
        // Storing Result In A List
        List<Integer> treeList = new LinkedList<>();
        
        // Calling In Order Function
        inorder(treeList, root);
        
        // Returning Final Result
        return treeList;
        
    }
    
    // Inorder Function
    public void inorder(List<Integer> list, TreeNode root){
        
        // In case The Root Is Null We Stop The Recursive Call
        // This Is The Base Condition
        if(root == null){
            return;
        }
        
        // We Traverse Through All The Left Nodes First
        inorder(list, root.left);
        
        // Then We Add The Values
        list.add(root.val);
        
        // And Finally We Traverse The Right Nodes Of The Tree.
        inorder(list, root.right);
        
    }
}

File: BinaryTreeInorderTraversal.js
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// TC: O(N) SC: O(N) Recursion
const traversal = (root) => {
    if(!root) return [];
    
    const node = root;
    const result = [];
    
    const inorder = (root) => {
        if(!root) return;
        if(root.left) inorder(root.left);
        result.push(root.val);
        if(root.right) inorder(root.right);
    }
    inorder(node);
    return result;
}

var inorderTraversal = function(root) {
    return traversal(root);
};



//TC: O(N) SC: O(N) Iteration
var inorderTraversal = function(root) {
    const result = [];
    const stack = [];
    let current = root;
    while(current || stack.length > 0) {
        while(current) {
            stack.push(current);
            current = current.left;
        }
            current = stack.pop();
            result.push(current.val);
            current = current.right;
    }
    return result;
};
